First draft

jeanconn opened the pull request:
First draft of a quick tool to look for image data SAR opportunities and select stars.

I wanted to use sparkles for the reporting and wanted to be able to add a couple of custom comment/message to an acar from the calling code, and couldn't figure out a handy way of doing that except for hacking _run_aca_review to *not* clear the messages.  For this application it probably isn't worth a change in sparkles (though might be useful for something else) to make a path to do that.  Or I just missed a way that already exists to do that.

jeanconn committed:
First draft

jeanconn committed:
Unwrap for easier debugging

jeanconn committed:
Handle no opportunities and add print info

jeanconn commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,264 @@
+        ncat, star1, star2 = sub_in_stars(cat)
+
+        # Use those stars in the three observations in the splits
+        cargs['include_ids_guide'] = list(ncat.guides['id'])
```
I think it makes the most sense to force include all 8 stars in all 3 observations (so that different temperatures and dither don't cause different star selection) but could also just check that the same 8 stars are selected.

taldcroft replied:
Agreed on forcing all stars after the initial catalog is selected.

jeanconn commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,264 @@
+    aca_arr = [acas[cat] for cat in acas]
+    aca_arr = sorted(aca_arr, key=lambda k: k.meta['date'])
+
+    # Do we want to parse the ms to check the last one?  Otherwise no duration.
+    candidates = []
+    for idx, aca in enumerate(aca_arr[0:-1]):
```
It might make sense for the ACA reviewer to just manually check the schedule summary to see if the last dwell is a long ER.

taldcroft replied:
Sounds fine.

taldcroft replied:
Always write `arr[:n]`, never `arr[0:n]`.

taldcroft replied:
This is better written with the standard idiom for iterating over neighbors in a list. This is a frequent idiom so try to memorize it.
```
for aca0, aca1 in zip(aca_arr[:-1], aca_arr[1:]):
```
Or
```
for aca, next_aca in ...:
```

jeanconn committed:
Just let sparkles version be sparkles version

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+import chandra_aca
+from sparkles.core import ACAReviewTable, get_summary_text, get_acas_from_pickle
+import proseco
+from Chandra.Time import DateTime
```
Do not use `DateTime` in new code unless there is a specific reason.

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+    return args
+
+
+def _run_aca_review(load_name=None, *, acars=None, make_html=True, report_dir=None,
```
Add a comment for future readers about why this is here and how it is different from the code in sparkles.

jeanconn replied:
Yep.  That was in my plan if you didn't figure out a better way to use sparkles for this.

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+    return naca, star1, star2
+
+
+def do(load_name):
```
What is this `do()` buying us here? I would just call it `main()` and factor out some low-hanging fruit like `get_candidates()`

jeanconn replied:
For me it was handy to separate the method calling get_options from the method doing the task.  But I'm probably fine now that there's a working draft.

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+def do(load_name):
+    acas = pickle.load(gzip.open(load_name))
+    aca_arr = [acas[cat] for cat in acas]
+    aca_arr = sorted(aca_arr, key=lambda k: k.meta['date'])
```
Everywhere replace `aca.meta['attr']` with `aca.attr`.

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+
+def do(load_name):
+    acas = pickle.load(gzip.open(load_name))
+    aca_arr = [acas[cat] for cat in acas]
```
Better written as `aca_arr = sorted(acas.values(), key=lambda aca: aca.date)`

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+        man_dur = duration(aca.meta['att'], next_aca.meta['att'])
+        man_start = DateTime(next_aca.meta['date']).secs - man_dur
+        obs_dur = man_start - DateTime(aca.meta['date']).secs
+        if (obsid > 40000) and (obsid < 59000) and (obs_dur >= MIN_DWELL):
```
```suggestion
        man_start = CxoTime(next_aca.date) - man_dur * u.s
        obs_dur = man_start - CxoTime(aca.date)
        if (obsid > 40000) and (obsid < 59000) and (obs_dur >= MIN_DWELL):  # MIN_DWELL = 13 * u.ks
```

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+        sar_args = [cargs.copy(), cargs.copy(), cargs.copy()]
+        sar_args[0]['obsid'] = obsid
+
+        sar_args[1]['date'] = DateTime(DateTime(sar_args[0]['date']).secs + 4250).date
```
```suggestion
        sar_start = CxoTime(sar_args[0]['date'])
        four_ks = 4.25 * u.ks
        sar_args[1]['date'] = (sar_start + four_ks).date
```

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+        sar_args[1]['dither_guide'] = (8, 8)
+        sar_args[1]['img_size_guide'] = 6
+
+        sar_args[2]['date'] = DateTime(DateTime(sar_args[0]['date']).secs + 8500).date
```
```suggestion
        sar_args[2]['date'] = (sar_start + 2 * four_ks).date
```

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+def sub_in_stars(aca):
+
+    # Sort the candidates by mag
+    aca.guides.cand_guides.sort('mag')
```
It's never a good idea to do an in-place munge of an argument that is passed in unless that is explicitly the goal of the function. I would do:
```
cand_guides = aca.guides.cand_guides.copy()
cand_guides.sort('mag', reverse=True)
```
For the rest you might do (not 100% sure if this is right but seems OK):
```
guide_ids = set(aca.guides['id'])
for star1, star2 in zip(cand_guides[:-1], cand_guides[1:]):
    if star1['id'] in guide_ids and star2['id'] in guide_ids:
        continue
```

jeanconn replied:
Yeah, I was admittedly lazy (but did recognize this was an in-place operation, on an object that isn't really used again with an original sort that also isn't used for anything).  And sure, I can walk the offset lists instead of taking them out in off-by-one order, to improve readability.

taldcroft commented on image_data_sar/sched.py:
```diff
@@ -0,0 +1,262 @@
+    args = aca.call_args.copy()
+
+    # Sub in the faintest and stop if it still has no criticals
+    for i, n in enumerate(range(len(aca.guides.cand_guides) - 1, len(aca.guides) - 1, -1)):
```
I'm not sure about cutting off at the number of guides since I think that assumes the guide stars are all the brightest.

jeanconn replied:
Good point!  I suppose I wasn't as onboard with this code to start so didn't think it through.  So, which stars should we check as candidates?  All the ones that are fainter than the faintest guide star in the original catalog?

taldcroft replied:
Also a good point to filter `cand_guides` on stars that are fainter than the faintest in the catalog. Then you don't need the `if star1['id'] in guide_ids and star2['id'] in guide_ids:` test and things become even simpler!

taldcroft commented:
Looks generally good.

taldcroft commented:
Have you tested this on cases where it actually finds some opportunities?

jeanconn commented:
Thanks for the feedback Tom!  Overall, I couldn't figure out what code really belongs in here and what I'd just do manually based on my review, so I hope some of this looks like it would be helpful to you in your prelim check!

jeanconn committed:
Zip over the acas

jeanconn committed:
Use attr access instead of meta dict

jeanconn committed:
Simplify making the array of acas

jeanconn committed:
Replace DateTime with CxoTime

jeanconn committed:
Simplify the selection of the faint ones

jeanconn committed:
Remove do() function.

With regard to why it was there, I was finding it useful to have
the function without options-handling available.  But maybe that
debugging is done.

jeanconn committed:
Comment on ending on ER with no candidates.

jeanconn committed:
Add comments about _run_aca_review

jeanconn committed:
Add docstring for subbing function

jeanconn committed:
Print more debug data for no opportunities
